src/main/scala/viper/server/vsi/Cache.scala

```diff
@@ -0,0 +1,204 @@
+package viper.server.vsi
+
+import viper.silver.utility.CacheHelper
+
+import scala.collection.mutable.{ListBuffer, Map => MutableMap}
+
+/** The goal of this generic caching trait is to provide
+  *
+  * 1) an elaborate storing system that can decide whether or not stored resources are valid.
+  *
+  * in order to implement
+  *
+  * 2) A very simple caching API
+  *
+  * A verification server should - very bluntly expressed - take a program and return a (potentially
+  * empty) set of errors. This process is assumed to be deterministic. I.e., verifying the same program twice
+  * results in the exact same verification result. Of course, changing the program is likely to change
+  * the verification outcome. However, it may well be that the outcome will not completely change. E.g.,
+  * if a program consists of two totally independent methods, changing the first may have absolutely
+  * no effect on the verification results of the second. In such a situation resources can be economized
+  * by reverifying only the first method and not reverifying the second. Sine one of the methods is not reverified
+  * its verifications (from a previous verification) must be stored in order to return them.
+  *
+  * As mentioned earlier, this requires an elaborate storage system that can
+  *
+  *   1) store verification results for individual program members
+  *   2) decide whether or not stored results are still valid if the program is changed.
+  *
+  * Once such a system is in place it can be used to implement a simple API. This API should take a
+  * program P, analyse it, transforms it and return a program P' as well as a set of verification results.
+  * This program should thereby have been transformed in such a way that a verifier may not perform
+  * unnecessary verifications on its members.
+  * */
+abstract class VerificationServerInterfaceCache {
+
+  override def toString: String = _cache.toString
+
+  /** The cache infrastructure is a nested, mutable Map. The maps can be described in terms of their
+    * key and values as follows:  (FileName -> (hashString -> cacheEntries)).
+    *
+    * The inner map is referred to as the fileCache. As the name indicates, it stores,
+    * for each file, a number of hashes and corresponding cache entries.
+    */
+  protected val _cache = MutableMap[String, FileCash]()
+  type FileCash = MutableMap[String, List[CacheEntry]]
+
+  /** This method transforms a program by applying cache's current state.
+    *
+    * The input and out put program should be equivalent with respect to verification, but they might
+    * differ in their AST. In particular the, output program should might be transformed such that
+    * verifying it is faster than verifying the input program.
+    *
+    * Additionally, the method returns previously cached verification result for members of the
+    * ast that hit the cache.
+    * */
+  def apply(
```

Comment:
Check that naming the method ```apply``` does not carry idiomatic semantics in Scala (as that would happen, e.g., inside a companion object).